api_app/analyzers_manager/observable_analyzers/maxmind.py

```diff
@@ -6,69 +6,93 @@
-                ):
-                    raise AnalyzerRunException(
-                        f"failed extraction of maxmind db {db},"
-                        " reached max number of attempts"
-                    )
-                if not os.path.exists(db_location):
-                    raise maxminddb.InvalidDatabaseError(
-                        "database location does not exist"
-                    )
-                reader = maxminddb.open_database(db_location)
-                maxmind_result = reader.get(self.observable_name)
-                reader.close()
-            except maxminddb.InvalidDatabaseError as e:
-                error_message = f"Invalid database error: {e}"
-                logger.exception(error_message)
-                maxmind_result = {"error": error_message}
-            logger.info(f"maxmind result: {maxmind_result}")
+        for db in self._supported_dbs:
+            maxmind_result = self._query_single_db(observable_query, db, api_key)
+
             if maxmind_result:
+                logger.info(f"maxmind result: {maxmind_result} in {db=}")
                 maxmind_final_result.update(maxmind_result)
             else:
-                logger.warning("maxmind result not available")
+                logger.warning(f"maxmind result not available in {db=}")
 
         return maxmind_final_result
 
     @classmethod
-    def _get_api_key(cls) -> Optional[str]:
-        for plugin in PluginConfig.objects.filter(
-            parameter__python_module=cls.python_module,
-            parameter__is_secret=True,
-            parameter__name="api_key_name",
-        ):
-            if plugin.value:
-                return plugin.value
-        return None
+    def _get_physical_location(cls, db: str) -> str:
+        return f"{settings.MEDIA_ROOT}/{db}{cls._default_db_extension}"
+
+    def _query_single_db(self, query_ip: str, db_name: str, api_key: str) -> dict:
+        result: ASN | City | Country
+        db_path: str = self._get_physical_location(db_name)
+        self._check_and_update_db(api_key, db_name)
+
+        logger.info(f"Query {db_name=} for {query_ip=}")
+        with Reader(db_path) as reader:
+            try:
+                if "ASN" in db_name:
+                    result = reader.asn(query_ip)
+                elif "Country" in db_name:
+                    result = reader.country(query_ip)
+                elif "City" in db_name:
+                    result = reader.city(query_ip)
+            except AddressNotFoundError:
+                reader.close()
+                logger.info(
+                    f"Query for observable '{query_ip}' "
+                    "didn't produce any results in any db."
+                )
+                return {}
+            except (GeoIP2Error, maxminddb.InvalidDatabaseError) as e:
+                error_message = f"GeoIP2 database error: {e}"
+                logger.exception(error_message)
+                return {"error": error_message}
```

Comment:
c'è proprio un modo specifico per inserire errori del framework:
```
            self.report.errors.append(error_message)
            self.report.save()
```

Reply:
e puoi ritornare sempre vuoto

Reply:
the problem with this approach is that i'm in `MaxmindDBManager` that is not a subclass of `ObservableAnalyzer`.
I tried to change the code accordingly.
Now the `query_all_dbs` method returns 2 dicts, one with the results and one with the errors.